### PR TITLE
[ome_application_certificate] Added certificate chain feature

### LINKED ISSUE #645: [BUG]: ome_application_certificate doesn't support uploading certificate chain
### Bug Description

When migrating from OME 3.10 to OME 4.x appliances, a requirement is that the full certificate chain is uploaded to the appliance instead of just the standard cert.
In the oem_appliacation_certificate module, even if the `upload_file` contains the full certificate chain, it is not imported as the full chain, and causes the migration to be blocked.

Another parameter should be added to import either the Certificate or Certificate Chain, with a default of Certificate
image

I lost about 3 hours of work generating and importing certs, only to find out that if I used the GUI to import the same certificate I was providing to the Ansible module and selected Certificate Chain, it worked as expected

### Component or Module Name

ome_application_certificate

### Ansible Version

2.16

### Python Version

3.10

### iDRAC/OME/OME-M version

OME 3.10.2 and OME 4.0.1

### Operating System

N/A

### Playbook Used

```yaml
    - name: Upload the signed certificate
      dellemc.openmanage.ome_application_certificate:
        hostname: "{{ inventory_hostname }}"
        username: "{{ dell_ome_username }}"
        password: "{{ dell_ome_password }}"
        validate_certs: "{{ dell_validate_certs | default(omit) }}"
        ca_path: "{{ dell_ca_path | default(omit) }}"
        command: "upload"
        upload_file: "/tmp/{{ inventory_hostname }}.cer"
      delegate_to: localhost
      tags:
        - upload
```

### Logs

N/A

### Steps to Reproduce

Create a .cer file that contains the full certificate chain and use the ansible module to upload the certificate chain file

Observe no errors and new cert is used in OME, but is not imported as the Certificate Chain, causing a validation error failure in the Migrate Out/In steps

### Expected Behavior

If a full chain is provided, it will be uplaoded as the full chain, or a parameter added to specify if the file being provided is a certificate or a certificate chain

### Actual Behavior

Observe no errors and new cert is used in OME, but is not imported as the Certificate Chain, causing a validation error failure in the Migrate Out/In steps

### Screenshots

image

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Saksham-Nautiyal**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### OUTPUT

```paste below

```
##### ADDITIONAL INFORMATION

```paste below

```
# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

### CODE REVIEW

**Comment by ansible-collections-svc** [comment]
Can one of the admins verify this patch?

**Comment by sachin-apa** [comment]
run e2e

File: plugins/modules/ome_application_certificate.py [hunk 1]
```diff
@@ -160,37 +171,60 @@
...
+    elif command == "upload_cert_chain":
+        file_path = module.params["upload_file"]
+        uri = csr_uri.format("UploadCertChain")
+        if os.path.exists(file_path):
+            with open(file_path, 'rb') as file:
+                payload = file.read()
         else:
             module.fail_json(msg="No such file or directory.")
+
+    else:
```
Comment (sachin-apa) [inline plugins/modules/ome_application_certificate.py]: How will code reach invalid command when it is a choices.

File: plugins/modules/ome_application_certificate.py [hunk 2]
```diff
@@ -160,37 +171,60 @@
...
-from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError
+from urllib.error import URLError, HTTPError
 from ansible.module_utils.urls import ConnectionError, SSLValidationError
+from ansible.module_utils.compat.version import LooseVersion
+
+OME_INFO = "ApplicationService/Info"
+LOW_OME = "OME version low"
 
 
 def get_resource_parameters(module):
```
Comment (sachin-apa) [inline plugins/modules/ome_application_certificate.py:210]: Can we refactor this method to remove if else statements.

File: plugins/modules/ome_application_certificate.py [hunk 3]
```diff
@@ -160,37 +171,60 @@
 import json
 import os
 from ansible_collections.dellemc.openmanage.plugins.module_utils.ome import RestOME, OmeAnsibleModule
-from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError
+from urllib.error import URLError, HTTPError
 from ansible.module_utils.urls import ConnectionError, SSLValidationError
+from ansible.module_utils.compat.version import LooseVersion
+
+OME_INFO = "ApplicationService/Info"
+LOW_OME = "OME version low"
```
Comment (sachin-apa) [inline plugins/modules/ome_application_certificate.py]:
```suggestion
LOW_OME = "This operation is not supported on OpenManage Enterprise version {version}"
```
{version} replace with current version.

File: plugins/modules/ome_application_certificate.py [hunk 4]
```diff
@@ -209,45 +243,61 @@ def format_csr_string(csr_string):
     return formatted_csr
 
 
+def get_ome_version(rest_obj):
```
Comment (sachin-apa) [inline plugins/modules/ome_application_certificate.py]: Check if there a util for this already, if not move this to utils.

File: plugins/modules/ome_application_certificate.py [hunk 5]
```diff
@@ -209,45 +243,61 @@ def format_csr_string(csr_string):
...
                     resp_copy = resp.json_data
-                    formated_csr = format_csr_string(resp_copy["CertificateData"])
-                    resp_copy["CertificateData"] = formated_csr
+                    formatted_csr = format_csr_string(resp_copy["CertificateData"])
+                    resp_copy["CertificateData"] = formatted_csr
                     module.exit_json(msg="Successfully generated certificate signing request.",
                                      csr_status=resp_copy)
-                module.exit_json(msg="Successfully uploaded application certificate.", changed=True)
+                else:
+                    module.exit_json(msg="Successfully uploaded application certificate.", changed=True)
```
Comment (sachin-apa) [inline plugins/modules/ome_application_certificate.py]: Move the success and error messages to constant. 
Please change wherever applicable

File: plugins/modules/ome_application_certificate.py [hunk 6]
```diff
@@ -62,14 +62,16 @@
     version_added: 8.1.0
   upload_file:
     type: str
-    description: Local path of the certificate file to be uploaded. This option is applicable for C(upload).
-        Once the certificate is uploaded, OpenManage Enterprise cannot be accessed for a few seconds.
+    description: Local path of the certificate file to be uploaded. This option is applicable for C(upload) and C(upload_cert_chain).
+        Once the certificate is uploaded, OpenManage Enterprise cannot be accessed for a few seconds. The formats can be in .crt, .cer,
```
Comment (rajshekarp87) [inline plugins/modules/ome_application_certificate.py]:
```suggestion
        Once the certificate is uploaded, OpenManage Enterprise cannot be accessed for a few seconds. The formats can be .crt, .cer,
```

File: plugins/modules/ome_application_certificate.py [hunk 7]
```diff
@@ -62,14 +62,16 @@
...
+        .ca-bundle, .p7b, .der, and .pem format.
```
Comment (rajshekarp87) [inline plugins/modules/ome_application_certificate.py]:
```suggestion
        .ca-bundle, .p7b, .der, and .pem.
```

File: plugins/modules/ome_application_certificate.py (hunk 3 above)
Comment (rajshekarp87) [inline plugins/modules/ome_application_certificate.py]:
```suggestion
UNSUPPORTED_OME = "OME version low"
```

File: plugins/modules/ome_application_certificate.py (hunk 1 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 2 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py:210]: Done

File: plugins/modules/ome_application_certificate.py (hunk 3 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 5 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 6 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 7 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 3 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 4 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py]: Done

**Review by sachin-apa — DISMISSED** [review]

**Review by ShivamSh3 — DISMISSED** [review]

File: plugins/module_utils/ome.py [hunk 8]
```diff
@@ -426,6 +427,10 @@ def get_job_execution_details(self, job_id):
             pass
         return job_detail_status
 
+    def get_ome_version(self):
+        resp = self.invoke_request('GET', OME_INFO)
```
Comment (ABHISHEK-SINHA10) [inline plugins/module_utils/ome.py]: Put this code in utils.py and not ome.py

File: plugins/modules/ome_application_certificate.py [hunk 9]
```diff
@@ -62,14 +65,16 @@
...
-        Once the certificate is uploaded, OpenManage Enterprise cannot be accessed for a few seconds.
+    description: Local path of the certificate file to be uploaded. This option is applicable for C(upload) and C(upload_cert_chain).
+      Once the certificate is uploaded, OpenManage Enterprise cannot be accessed for a few seconds. The formats of the certificate
+      file are .crt, .cer, .ca-bundle, .p7b, .der, and .pem.
 requirements:
     - "python >= 3.9.6"
 author:
   - "Felix Stephen (@felixs88)"
   - "Kritika Bhateja (@Kritika-Bhateja-03)"
   - "Jennifer John (@Jennifer-John)"
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/ome_application_certificate.py:76]: Add me as an author too, I forgot last time, when added format_csr_string function.

- "Abhishek Sinha (@ABHISHEK-SINHA10)"

File: tests/unit/plugins/modules/test_ome_application_certificate.py [hunk 10]
```diff
@@ -105,6 +106,26 @@ def test_upload_csr_success(self, mocker, ome_default_args, ome_connection_mock_
...
+        assert exc.value.args[0] == "No such file or directory."
+
+    def test_upload_cert_chain_success(self, mocker, ome_default_args, ome_connection_mock_for_application_certificate,
+                                       ome_response_mock):
+        payload = "--BEGIN-REQUEST--"
+        mocker.patch(MODULE_PATH + 'ome_application_certificate.get_resource_parameters',
+                     return_value=("POST", "ApplicationService/Actions/ApplicationService.UploadCertChain", payload))
+        mocker.patch(MODULE_UTILS_PATH + "RestOME.get_ome_version", return_value="4.1.0")
+        ome_default_args.update({"command": "upload_cert_chain", "upload_file": "/path/certificate_chain.cer"})
+        ome_response_mock.success = True
```
Comment (ABHISHEK-SINHA10) [inline tests/unit/plugins/modules/test_ome_application_certificate.py]: In success UT, make sure to assert the success msg, assertion is not right.

Comment (Saksham-Nautiyal) [reply tests/unit/plugins/modules/test_ome_application_certificate.py]: Done

File: plugins/module_utils/ome.py (hunk 8 above)
Comment (Saksham-Nautiyal) [reply plugins/module_utils/ome.py]: Done

File: plugins/modules/ome_application_certificate.py (hunk 9 above)
Comment (Saksham-Nautiyal) [reply plugins/modules/ome_application_certificate.py:76]: Done

**Review by ABHISHEK-SINHA10 — DISMISSED** [review]

**Review by ABHISHEK-SINHA10 — APPROVED** [review]

**Review by rajshekarp87 — APPROVED** [review]

**Pull request merged**